### PR TITLE
Feat: copy programme membership table TIS21-2509

### PR DESCRIPTION
**Pull request by ReubenRobertsHEE**
This duplicates the existing ProgrammeMembership table into CurriculumMembership. This will prevent issues with partial programmeMembership data being saved into TraineeProfiles pending a complete sync of the new CurriculumMembership table from TCS.

There is some weirdness with Mongock not picking up the migration-scan-package when used via @EnableMongock and requiring manual instantiation on some machines.

TIS21-2509: ProgrammeMembership to CurriculumMembership - TSS Sync

### CODE REVIEW

File: src/test/java/uk/nhs/hee/tis/trainee/sync/migration/CopyProgrammeMembershipToCurriculumMembershipTest.java
Comment (ReubenRobertsHEE) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/migration/CopyProgrammeMembershipToCurriculumMembershipTest.java:66]: This feels a bit like the test is knowing too much about the inner workings of the method under test, but I'm not sure how else to establish reasonable coverage at this point.

File: src/test/java/uk/nhs/hee/tis/trainee/sync/migration/CopyProgrammeMembershipToCurriculumMembershipTest.java
Comment (ReubenRobertsHEE) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/migration/CopyProgrammeMembershipToCurriculumMembershipTest.java:100]: This feels a **_lot_** like the test is knowing too much about the inner workings of the method under test, but I'm not sure how else to establish reasonable coverage at this point.

**Comment by Judge40** [comment]
We may just have to put up with the tight coupling in this case, but I'd like to have a proper look over it myself before merging.

**Review by Judge40 — DISMISSED** [review]
I share the concerns about the tight coupling, creating real instances of the Mongo classes is difficult because the constructors are parameter heavy (with more and more Mongo types).

We're definitely on the edge of unit test territory here, an integration test would be much more suitable but we don't yet have a solution implemented for an in-memory (or other test accessible) mongo instance.

**Comment by ReubenRobertsHEE** [comment]
@hihilary before I merge (I accidentally fixed some code smells after Andy's review, so I'll need a fresh approval anyway) - would you mind pulling and seeing if the mongock migration runs for you with just the annotation (i.e. as the code stands) - or if you need to uncomment the @Bean and remove the @EnableMongock annotation? That might help get to the bottom of why there's been this issue 😃

**Comment by hihilary** [comment]
I copied the "ProgrammeMembership" collection from preprod into my local, and tried the data migration again. Both of @EnableMongock and @bean methods work on my side. 
@ReubenRobertsHEE have you tried a empty collection of "ProgrammeMembership"? I'm wondering if you get the same error as me and if some null check needs to be added.

**Comment by ReubenRobertsHEE** [comment]
> I copied the "ProgrammeMembership" collection from preprod into my local, and tried the data migration again. Both of @EnableMongock and @bean methods work on my side. @ReubenRobertsHEE have you tried a empty collection of "ProgrammeMembership"? I'm wondering if you get the same error as me and if some empty value check needs to be added.

Thanks Yafang, good to know both ways work on your side. I'll add an empty collection check - shouldn't happen in real life (given the code will be run only once), but better safe than sorry 😁

**Comment by ReubenRobertsHEE** [comment]
I've also removed the commented-out bean creation. It must be something weird on my side.

**Review by hihilary — APPROVED** [review]

**Pull request merged**
